### PR TITLE
docs: add guide for releasing an app with code push

### PR DESCRIPTION
**Pull request by bryanoltman**
## Description

Adds a walkthrough of a real Play Store release and applying a patch to it, using Time Shift as a demo.

Resolves https://github.com/shorebirdtech/docs/issues/25

## Type of Change

- [ ] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [x] 📝 Documentation
- [ ] 🗑️ Chore

### CODE REVIEW

File: docs/guides/code_push_release.md [hunk 1]
```diff
@@ -0,0 +1,252 @@
+---
+sidebar_position: 2
+title: 🚢 Release a Code Push App
+description: Release a Code Push App
+---
+
+# Releasing an code push app
```
Comment (felangel) [inline docs/guides/code_push_release.md:7]:
```suggestion
# Releasing a code push app
```

File: docs/guides/code_push_release.md [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+
+This guide walks you through releasing a code push app to the Play Store and applying a patch to that release.
+
+## Prerequisites
+
+This guide assumes that you have an existing Shorebird app. If you don't have one, you can create one by following the [code push quickstart](./code_push_quickstart/) guide.
+
+The app we will be releasing in this guide is [`Time Shift`](https://play.google.com/store/apps/details?id=dev.shorebird.u_shorebird_clock), our demo code push app. ([source](https://github.com/shorebirdtech/time_shift/))
+
+## Creating a Release
```
Comment (felangel) [inline docs/guides/code_push_release.md:17]:
```suggestion
## Creating a release
```

nit but can we standardize the casing?

File: docs/guides/code_push_release.md [hunk 3]
```diff
@@ -0,0 +1,252 @@
+---
+sidebar_position: 2
+title: 🚢 Release a Code Push App
+description: Release a Code Push App
+---
+
+# Releasing an code push app
+
+This guide walks you through releasing a code push app to the Play Store and applying a patch to that release.
```
Comment (felangel) [inline docs/guides/code_push_release.md:9]:
```suggestion
This guide walks through releasing a code push app to the Play Store and applying a patch to that release.
```

File: docs/guides/code_push_release.md [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+├─────────┼──────┤
+│ 1.0.1   │ --   │
+├─────────┼──────┤
+│ 1.0.2+1 │ --   │
+├─────────┼──────┤
+│ 1.0.2+5 │ --   │
+└─────────┴──────┘
+```
+
+From this, we can see that the most recent release is `1.0.2+5`. This corresponds with what we see in the Play Store console:
```
Comment (felangel) [inline docs/guides/code_push_release.md:38]:
```suggestion
We can see that the most recent release is `1.0.2+5`. This corresponds with what we see in the Play Store console:
```

File: docs/guides/code_push_release.md [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+
+This guide assumes that you have an existing Shorebird app. If you don't have one, you can create one by following the [code push quickstart](./code_push_quickstart/) guide.
+
+The app we will be releasing in this guide is [`Time Shift`](https://play.google.com/store/apps/details?id=dev.shorebird.u_shorebird_clock), our demo code push app. ([source](https://github.com/shorebirdtech/time_shift/))
+
+## Creating a Release
+
+### Determine next release version
+
+We will start by running `shorebird releases list` to see the current set of releases:
```
Comment (felangel) [inline docs/guides/code_push_release.md:21]:
```suggestion
We will start by running `shorebird releases list` from the root directory of our project to see the current set of releases:
```

File: docs/guides/code_push_release.md [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+```diff
+name: time_shift
+description: Demo app showing Shorebird updates.
+publish_to: "none"
+
+-version: 1.0.2+5
++version: 1.0.3+6
+
+environment:
+  sdk: ">=2.19.4 <3.0.0"
```
Comment (felangel) [inline docs/guides/code_push_release.md:83]: side-note we should probably upgrade to Dart 3 😄

File: docs/guides/code_push_release.md [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+├─────────┼──────┤
+│ 1.0.2+5 │ --   │
+└─────────┴──────┘
+```
+
+From this, we can see that the most recent release is `1.0.2+5`. This corresponds with what we see in the Play Store console:
+
+![ReleaseVersion](https://github.com/shorebirdtech/docs/assets/581764/e6b6c276-49de-4142-8f32-dbf5e41379fa)
+
+The release we are going to create is `1.0.3+6`.
```
Comment (felangel) [inline docs/guides/code_push_release.md:42]:
```suggestion
Next, we'll create a new release for version `1.0.3+6`.
```

File: docs/guides/code_push_release.md [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+
+### Make code changes
+
+Version `1.0.3+6` of Time Shift will be a small patch that changes the default clock face from `particle` to `generative`.
```
Comment (felangel) [inline docs/guides/code_push_release.md:46]:
```suggestion
Version `1.0.3+6` of Time Shift will consist of a small patch that changes the default clock face from `particle` to `generative`.
```

File: docs/guides/code_push_release.md [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+
+```diff
+   final clock = ClockFace.values.firstWhere(
+     (clock) => clock.name == clockName,
+-    orElse: () => ClockFace.particle,
++    orElse: () => ClockFace.generative,
+   );
+```
+
+We run the app using `shorebird run` to make sure this change does what we expect.
```
Comment (felangel) [inline docs/guides/code_push_release.md:58]:
```suggestion
Verify the change does what we expect using `shorebird run`.
```

File: docs/guides/code_push_release.md [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+
+After verifying, we will commit this change and push it to GitHub:
```
Comment (felangel) [inline docs/guides/code_push_release.md:60]:
```suggestion
Once the change has been verified, commit this change and push it to GitHub:
```

File: docs/guides/code_push_release.md [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+git add lib/main.dart
+git commit -m "Change default clock face to generative"
+git push
+```
+
+[On GitHub](https://github.com/shorebirdtech/time_shift/commit/d1fe9451aa18a775163bce95dd9dab551aaf6259)
+
+### Update app version
+
+We will now update the app version in `pubspec.yaml`:
```
Comment (felangel) [inline docs/guides/code_push_release.md:72]:
```suggestion
Next, bump the app version in the `pubspec.yaml`:
```

File: docs/guides/code_push_release.md [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+publish_to: "none"
+
+-version: 1.0.2+5
++version: 1.0.3+6
+
+environment:
+  sdk: ">=2.19.4 <3.0.0"
+```
+
+We will commit this change, tag it, and push it.
```
Comment (felangel) [inline docs/guides/code_push_release.md:86]:
```suggestion
Commit, tag, and push the change.
```

File: docs/guides/code_push_release.md [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+Your next step is to upload the app bundle to the Play Store.
+./build/app/outputs/bundle/release/app-release.aab
+
+See the following link for more information:
+https://support.google.com/googleplay/android-developer/answer/9859152?hl=en
+```
+
+### Upload to the Play Store
+
+As per the instructions above, we need to upload the generated `.aab` to the Play Store.
```
Comment (felangel) [inline docs/guides/code_push_release.md:132]:
```suggestion
As per the instructions above, we must upload the generated `.aab` to the Play Store.
```

File: docs/guides/code_push_release.md [hunk 14]
```diff
@@ -0,0 +1,252 @@
...
+
+1. Navigate to the [Play Console](https://play.google.com/console/developers).
+1. Choose your developer account (for us, it's Shorebirdbird.dev).
+1. Select the Time Shift app.
+1. Select "Testing -> Open Testing" from the side bar.
+1. Click the "Create new release" button.
+
+![CreateNewRelease](https://github.com/shorebirdtech/docs/assets/581764/90c9c7ed-bc39-4731-bfec-524f89e2baf6)
+
+You will now be prompted to upload the `.aab` file. You can find this file in `./build/app/outputs/bundle/release/app-release.aab`.
```
Comment (felangel) [inline docs/guides/code_push_release.md:142]:
```suggestion
Upload the `.aab` file located in `./build/app/outputs/bundle/release/app-release.aab`.
```

File: docs/guides/code_push_release.md [hunk 15]
```diff
@@ -0,0 +1,252 @@
...
+
+1. From `~/Shorebird/time_shift`, run `open ./build/app/outputs/bundle/release/` to open the folder containing the `.aab` in Finder.
```
Comment (felangel) [inline docs/guides/code_push_release.md:144]:
```suggestion
1. From the root directory of the project, run `open ./build/app/outputs/bundle/release/` to open the folder containing the `.aab` in Finder.
```

File: docs/guides/code_push_release.md [hunk 16]
```diff
@@ -0,0 +1,252 @@
...
+
+Click "Next" and then "Save" (both in the bottom-right corner) to submit.
+
+This will take you to the publishing overview page. Click "Submit for review" to submit the release for review.
+
+We should now see a release in the Play Store console with an "In review" status:
+
+![InReview](https://github.com/shorebirdtech/docs/assets/581764/4cfdc7fb-2049-4110-b1cd-da99c7a491f7)
+
+We will now wait for the Play Store to approve this release.
```
Comment (felangel) [inline docs/guides/code_push_release.md:159]:
```suggestion
Now you must wait for the Play Store to approve the release.
```

File: docs/guides/code_push_release.md [hunk 17]
```diff
@@ -0,0 +1,252 @@
...
+
+### Create a GitHub Release
+
+We also want to create a GitHub release. For us, that looks like:
```
Comment (felangel) [inline docs/guides/code_push_release.md:163]:
```suggestion
It's recommended to to create a GitHub release as well:
```

File: docs/guides/code_push_release.md (hunk 2 above)
Comment (bryanoltman) [reply docs/guides/code_push_release.md]: Done.

File: docs/guides/code_push_release.md [hunk 18]
```diff
@@ -0,0 +1,252 @@
...
+1. Add a description of the release ("Changes the default clock face to 'generative'").
+1. Publish the release.
+
+## After the Release is Approved
+
+Once the release has been approved, you will be able to download it from the Play Store.
+
+## Creating a Patch
+
+Let's say we want to create a patch for `1.0.3+6` that fixes a bug.
```
Comment (felangel) [inline docs/guides/code_push_release.md:178]:
```suggestion
Patches can be pushed to fix bugs in the `1.0.3+6` without requiring a new submission to the Play Store.
```

File: docs/guides/code_push_release.md [hunk 19]
```diff
@@ -0,0 +1,252 @@
...
+
+### Make the change
+
+We start by checking out the `v1.0.3+6` release tag:
```
Comment (felangel) [inline docs/guides/code_push_release.md:182]:
```suggestion
Start by checking out the `v1.0.3+6` release tag:
```

File: docs/guides/code_push_release.md [hunk 20]
```diff
@@ -0,0 +1,252 @@
...
+
+```sh
+git checkout v1.0.3+6
+```
+
+We will branch from here, as this change will represent a divergence from the `main` branch:
```
Comment (felangel) [inline docs/guides/code_push_release.md:188]:
```suggestion
Next create a branch, as this change will represent a divergence from the `main` branch:
```

File: docs/guides/code_push_release.md [hunk 21]
```diff
@@ -0,0 +1,252 @@
...
+For the purposes of this guide, we will change the default clock face back to `particle` in `lib/main.dart`:
+
+```diff
+     (clock) => clock.name == clockName,
+-    orElse: () => ClockFace.generative,
++    orElse: () => ClockFace.particle,
+   );
+```
+
+Commit this change and push the branch:
```
Comment (felangel) [inline docs/guides/code_push_release.md:203]:
```suggestion
Commit the change and push to the branch:
```

File: docs/guides/code_push_release.md [hunk 22]
```diff
@@ -0,0 +1,252 @@
...
+
+```sh
+git add lib/main.dart
+git commit -m "Change default clock face to particle"
+git push --set-upstream origin v1.0.3+6-patch1
+```
+
+We will also tag this commit as `v1.0.3+6-patch1`:
```
Comment (felangel) [inline docs/guides/code_push_release.md:211]:
```suggestion
Tag this commit as `v1.0.3+6-patch1`:
```

File: docs/guides/code_push_release.md [hunk 23]
```diff
@@ -0,0 +1,252 @@
...
+Would you like to continue? (y/N) Yes
+✓ Creating patch (0.1s)
+✓ Uploading artifacts (0.9s)
+✓ Fetching channels (90ms)
+✓ Promoting patch to stable (61ms)
+
+✅ Published Patch!
+```
+
+This change will now be available to users with version `1.0.3+6` of the app.
```
Comment (felangel) [inline docs/guides/code_push_release.md:252]:
```suggestion
This patch will now be available to users with version `1.0.3+6` of the app.
```

File: docs/guides/code_push_release.md (hunk 6 above)
Comment (bryanoltman) [reply docs/guides/code_push_release.md:83]: Agreed. Filed https://github.com/shorebirdtech/time_shift/issues/9.

File: docs/guides/code_push_release.md [hunk 24]
```diff
@@ -0,0 +1,252 @@
...
+Start by checking out the `v1.0.3+6` release tag:
+
+```sh
+git checkout v1.0.3+6
+```
+
+Next, create a branch, as this change will represent a divergence from the `main` branch:
+
+```sh
+git checkout -b v1.0.3+6-patch1
```
Comment (felangel) [inline docs/guides/code_push_release.md:191]: with this approach we'll just end up with lots of branches right? Couldn't we also delete the branch after the patch has been tagged?

Comment (bryanoltman) [reply docs/guides/code_push_release.md:191]: Yeah, I could see this getting a bit messy. Added cleanup instructions.

**Review by felangel — APPROVED** [review]

**Pull request merged**